x/qgb/abci.go

```diff
@@ -15,21 +15,51 @@ const SignificantPowerDifferenceThreshold = 0.05
 
 // EndBlocker is called at the end of every block.
 func EndBlocker(ctx sdk.Context, k keeper.Keeper) {
-	handleDataCommitmentRequest(ctx, k)
+	// we always want to create the valset at first so that if there is a new validator set, then it is
+	// the one responsible for signing from now on.
 	handleValsetRequest(ctx, k)
+	handleDataCommitmentRequest(ctx, k)
 }
 
 func handleDataCommitmentRequest(ctx sdk.Context, k keeper.Keeper) {
-	if ctx.BlockHeight() != 0 && ctx.BlockHeight()%int64(k.GetDataCommitmentWindowParam(ctx)) == 0 {
-		dataCommitment, err := k.GetCurrentDataCommitment(ctx)
+	setDataCommitmentAttestation := func() {
+		dataCommitment, err := k.NextDataCommitment(ctx)
 		if err != nil {
-			panic(sdkerrors.Wrap(err, "coudln't get current data commitment"))
+			panic(sdkerrors.Wrap(err, "couldn't get current data commitment"))
 		}
 		err = k.SetAttestationRequest(ctx, &dataCommitment)
 		if err != nil {
 			panic(err)
 		}
 	}
+	// this will  keep executing until all the needed data commitments are created and we catchup to the current height
+	for {
+		hasLastDataCommitment, err := k.HasDataCommitmentInStore(ctx)
+		if err != nil {
+			panic(err)
+		}
+		if hasLastDataCommitment {
+			// if the store already has a data commitment, we use it to check if we need to create a new data commitment
+			lastDataCommitment, err := k.GetLastDataCommitment(ctx)
+			if err != nil {
+				panic(err)
+			}
+			if ctx.BlockHeight()-int64(lastDataCommitment.EndBlock) >= int64(k.GetDataCommitmentWindowParam(ctx)) {
+				setDataCommitmentAttestation()
+			} else {
+				// the needed data commitments are already created and we need to wait for the next window to elapse
+				break
+			}
+		} else {
+			// if the store doesn't have a data commitment, we check if the window has passed to create a new data commitment
+			if ctx.BlockHeight() >= int64(k.GetDataCommitmentWindowParam(ctx)) {
+				setDataCommitmentAttestation()
+			} else {
+				// the first data commitment window hasn't elapsed yet to create a commitment
+				break
+			}
+		}
+	}
```

Comment:
> but do you mind adding a quick comment on line 371 where we're testing the loop functionality?

I guess the line number is incorrect

>  Also I think I missed it but are we testing the second exit condition?
I think it is tested implicitly by specifying the nonce. This means that no data commitment was created in that range

> I think there's a very minor gas optimization where we could load the data commitment window once instead of reading from state each loop. I know we're not the ones paying gas since its in endblock, and that shouldn't be breaking, so tbh its probably not even worth changing until we actually test the diff.

I will create a separate issue for this. But, who pays for end blocker? There are no transactions for the QGB module... So you mean the validators will pay for it?

Reply:
> But, who pays for end blocker?

no sorry, I meant that since it's requried to come to consensus, no one is paying, just like everything in EndBlock. Technically tho, we are reading from the state fewer times, so we're saving "gas" or simply node/network resources :slightly_smiling_face: 

> I guess the line number is incorrect

do you mind pointing our where in the test we are hitting that looping functionlity then? that still looks that the point in the test afaict :slightly_smiling_face:

https://github.com/SweeXordious/celestia-app/blob/3fc91f06187d2041d6fc6208fccecb3d5d2c95b1/x/qgb/abci_test.go#L371-L373

Reply:
> so we're saving "gas" or simply node/network resources slightly_smiling_face

I think that that simple statements like those will not affect it much. But yes, good point, will open an issue to optimize that.

For the test thing, the catchup should start here, but good point we're not testing all the possibilities. I will open an issue to do that.

I don't want to commit anymore to this PR not to wait for another round of  reviews.

Reply:
makes sense, lets :shipit: